src/textarea/Textarea.tsx

```diff
@@ -0,0 +1,250 @@
+  ) => void
+}
+
+type errorField = string | JSX.Element
+
+export interface TextAreaProps extends CommonFormFields {
+  defaultValue?: string
+  labelledBy?: string
+  onChange?: (obj: OnChangeParameters) => void
+  className?: Classcat.Class
+  error?: errorField
+  label?: string
+  buttonTitle?: string
+  focus?: boolean
+  pattern?: string
+  fieldRef?: (textarea: HTMLTextAreaElement) => void
+  focusBorder?: boolean
+  onButtonClick?: (event: React.MouseEvent<HTMLElement>) => void
+  buttonIcon?: JSX.Element
+  fitContent?: boolean
+}
+
+interface TextAreaAttributes extends CommonFormFields {
+  className?: string
+  value: string
+  ['aria-invalid']?: 'true' | 'false'
+  ['aria-labelledby']?: string
+  ref?: (textarea: HTMLTextAreaElement) => void
+  onChange?: (event: React.ChangeEvent<HTMLTextAreaElement>) => void
+}
+
+export interface TextAreaState {
+  readonly value: string
+  readonly hasFocus: boolean
+}
+
+const DisplayError = (error: errorField) => {
+  const className = 'kirk-error-message'
+
+  return React.isValidElement(error) ? (
+    React.cloneElement(error, { className } as Object)
+  ) : (
+    <span role="alert" className={className}>
+      {error}
+    </span>
+  )
+}
+
+export default class Textarea extends PureComponent<TextAreaProps, TextAreaState> {
+  private textarea: HTMLInputElement | HTMLTextAreaElement
+  static defaultProps: Partial<TextAreaProps> = {
+    fieldRef() {},
+    onFocus() {},
+    onBlur() {},
+    focusBorder: true,
+  }
+
+  state = {
+    value: this.props.defaultValue,
+    hasFocus: false,
+  }
+
+  buttonRef: HTMLButtonElement = null
+  textareaRef: RefObject<HTMLTextAreaElement> = React.createRef()
+
+  componentDidMount() {
+    if (this.textarea && this.props.focus) {
+      this.textarea.focus()
+    }
+  }
+
+  componentWillReceiveProps({ defaultValue, focus }: TextAreaProps) {
+    if (this.props.defaultValue !== defaultValue) {
+      this.setState({
+        value: defaultValue,
+      })
+    }
+    if (focus && this.props.focus !== focus) {
+      this.textarea.focus()
+    }
+  }
+
+  onFocus = (event: React.FocusEvent<HTMLTextAreaElement>) => {
+    this.setState({
+      hasFocus: true,
+    })
+    this.props.onFocus(event)
+  }
+
+  onBlur = (event: React.FocusEvent<HTMLTextAreaElement>) => {
+    if (!event.relatedTarget || event.relatedTarget !== this.buttonRef) {
+      this.setState({
+        hasFocus: false,
+      })
+      this.props.onBlur(event)
+    }
+  }
+
+  onTextAreaChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
+    this.setState(
+        {
+          value: event.target.value,
+        },
+        this.onChange,
+    )
+  }
+
+  onChange = () => {
+    this.props.onChange({
+      name: this.props.name,
+      value: this.state.value,
+    })
+
+    // Fit to content.
+    if (this.props.fitContent && this.textareaRef && this.textareaRef.current) {
+      this.textareaRef.current.style.height = '0'
```

Comment:
Why do you need to reset the height before updating it ?

Reply:
Then the scrollHeight is the height of the whole (overflowing) content.